### PR TITLE
THOR-1043 Non-Primitive Options

### PR DESCRIPTION
**Pull request by arciisine**
Everything should be up and running, and app tests have been resolved.

### CODE REVIEW

File: src/app/components/options-list/options-list.component.html [hunk 1]
```diff
@@ -35,6 +35,22 @@
...
+    <div mat-fill class="editor" *ngSwitchCase="'NON_PRIMITIVE'">
+        <label>{{ getOption(bindingKey).prettyName }}</label>
+        <ngx-monaco-editor #editor [options]="{
+                theme: 'vs', 
+                language: 'yaml', 
+                automaticLayout: true, 
+                lineNumbers: true,
+                minimap: {
+                    enabled: false
+                }
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/components/options-list/options-list.component.html:48]: Would these options be better saved as a variable in the .ts file?

File: src/app/widget-option.spec.ts [hunk 2]
```diff
@@ -476,3 +478,46 @@ describe('WidgetOptionCollection with bindings and custom fields', () => {
...
+    it('Objects update properly', () => {
+        const option = new WidgetNonPrimitiveOption('test', 'Test', {}, true);
+        expect(option.valueDefault).toEqual({});
+        option.intermediateValue = 'a: [1,2,3]';
+        expect(option.getValueToSaveInBindings()).toEqual({ a: [1, 2, 3] });
+        option.intermediateValue = `
+a: 5
+b: 10
+c: [1,2,3]
+`;
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/widget-option.spec.ts]: The indentation here annoys me, but I understand why it is necessary.  Is there a commonly-accepted standard for dealing with situations like this?  Would it be "cleaner" to use `\n` instead?

File: src/app/widget-option.spec.ts [hunk 3]
```diff
@@ -476,3 +478,46 @@ describe('WidgetOptionCollection with bindings and custom fields', () => {
         expect(options.testCustomFieldArray).toEqual([DatasetServiceMock.NAME_FIELD, DatasetServiceMock.TYPE_FIELD]);
     }));
 });
+
+describe('NonPrimitive Fields', () => {
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/widget-option.spec.ts:482]: I think you should add a test on setting `intermediateValue` to an improperly-formatted string and checking that it returns `undefined`.

File: src/app/widget-option.ts [hunk 4]
```diff
@@ -193,6 +194,9 @@ export class WidgetMultipleSelectOption extends WidgetOption {
 }
 
 export class WidgetNonPrimitiveOption extends WidgetOption {
+
+    private _intermediateValue: string;
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/widget-option.ts:198]: I like the `_intermediateValue` here.  I've added a note that we could likely use this strategy on validating numbers in the future too:  https://nextcentury.atlassian.net/browse/THOR-1044

**Review by ThomasSchellenbergNextCentury — CHANGES_REQUESTED** [review]
Please see specific comments for details.

File: src/app/components/options-list/options-list.component.html (hunk 1 above)
Comment (arciisine) [reply src/app/components/options-list/options-list.component.html:48]: Potentially, they are intimately tied to the component.  I view them akin to html attributes, vs external config.

File: src/app/widget-option.spec.ts (hunk 2 above)
Comment (arciisine) [reply src/app/widget-option.spec.ts]: Yaml will consume all the leading spaces of equal width, it just may not be apparent that is happening when comparing the values.  We could also build a function to strip the leading spaces consistently.

File: src/app/widget-option.spec.ts (hunk 3 above)
Comment (arciisine) [reply src/app/widget-option.spec.ts:482]: Sure, will add.

File: src/app/widget-option.spec.ts (hunk 2 above)
Comment (arciisine) [reply src/app/widget-option.spec.ts]: Relied on yaml stripping indentation

**Review by ThomasSchellenbergNextCentury — APPROVED** [review]

**Review by rartiss55 — APPROVED** [review]

**Review by Konnoke — APPROVED** [review]

**Pull request merged**